frame/bonded-finance/src/lib.rs

```diff
@@ -0,0 +1,383 @@
+		OptionQuery,
+	>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Create a new offer.
+		///
+		/// The dispatch origin for this call must be _Signed_ and the sender must have the
+		/// appropriate funds.
+		///
+		/// Emits a `NewOffer`.
+		#[pallet::weight(10_000)]
+		pub fn offer(origin: OriginFor<T>, offer: BondOfferOf<T>) -> DispatchResult {
+			let from = ensure_signed(origin)?;
+			Self::do_offer(&from, offer)?;
+			Ok(())
+		}
+
+		/// Bond to an offer.
+		///
+		/// The dispatch origin for this call must be _Signed_ and the sender must have the
+		/// appropriate funds.
+		///
+		/// Emits a `NewBond`.
+		#[pallet::weight(10_000)]
+		pub fn bond(
+			origin: OriginFor<T>,
+			offer: T::BondOfferId,
+			contracts: BalanceOf<T>,
+		) -> DispatchResult {
+			let from = ensure_signed(origin)?;
+			Self::do_bond(offer, &from, contracts)?;
+			Ok(())
+		}
+
+		/// Cancel an offer.
+		///
+		/// The dispatch origin for this call must be _Signed_ and the sender must be `AdminOrigin`
+		///
+		/// Emits a `OfferCancelled`.
+		#[pallet::weight(10_000)]
+		pub fn cancel_offer(origin: OriginFor<T>, offer: T::BondOfferId) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+			Self::do_cancel_offer(offer)
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		pub fn account_id(offer: T::BondOfferId) -> AccountIdOf<T> {
+			T::PalletId::get().into_sub_account(offer)
+		}
+
+		pub fn do_cancel_offer(offer_id: T::BondOfferId) -> DispatchResult {
+			let (creator, _) =
+				BondOffers::<T>::try_get(offer_id).map_err(|_| Error::<T>::BondOfferNotFound)?;
+			let offer_account = Self::account_id(offer_id);
+			T::NativeCurrency::transfer(&offer_account, &creator, T::Stake::get(), true)?;
+			BondOffers::<T>::remove(offer_id);
+			Self::deposit_event(Event::<T>::OfferCancelled { offer: offer_id });
+			Ok(())
+		}
+
+		#[transactional]
+		pub fn do_offer(
+			from: &AccountIdOf<T>,
+			offer: BondOfferOf<T>,
+		) -> Result<T::BondOfferId, DispatchError> {
+			ensure!(
+				offer.valid(
+					<T::Vesting as VestedTransfer>::MinVestedTransfer::get(),
+					T::MinReward::get()
+				),
+				Error::<T>::InvalidBondOffer
+			);
+			let offer_id = BondOfferCount::<T>::mutate(|offer_id| {
+				*offer_id = offer_id.next();
+				*offer_id
+			});
+			let offer_account = Self::account_id(offer_id);
+			T::NativeCurrency::transfer(from, &offer_account, T::Stake::get(), true)?;
+			T::Currency::transfer(
+				offer.reward_asset,
+				from,
+				&offer_account,
+				offer.reward_amount,
+				true,
+			)?;
+			BondOffers::<T>::insert(offer_id, (from.clone(), offer));
+			Self::deposit_event(Event::<T>::NewOffer { offer: offer_id });
+			Ok(offer_id)
+		}
+
+		#[transactional]
+		pub fn do_bond(
+			offer_id: T::BondOfferId,
+			from: &AccountIdOf<T>,
+			contracts: BalanceOf<T>,
+		) -> Result<BalanceOf<T>, DispatchError> {
+			BondOffers::<T>::try_mutate(offer_id, |offer| {
+				offer
+					.as_mut()
+					.map(|(creator, offer)| {
```

Comment:
this seems a bit weirdly aligned, I'd do a match and early exit to make it a bit clearer, instead of having the None condition at the end.

Reply:
:+1: